src/transformers/trainer.py

```diff
@@ -1062,6 +1074,9 @@ def train(
             self.model_wrapped = self.model
             gc.collect()  # force memory release
 
+        self.mem.stop("train")
+        self.mem.update_metrics("train", metrics)
```

Comment:
so the only thing I'm not sure how to handle is the conflict between the `on_log` callback needing to happen before the `on_train` (or eval, predict) but it's too soon to take a snapshot of memory metrics. They have to happen just before return.

So for now each stage returns the memory metrics, but they aren't passed to `log()`. Ideas how to overcome this chicken and egg problem?

And if I do move the log stage to last (which I don't think I can) there is a bit of a test problem in `test_trainer.py`'s `check_trainer_state_are_the_same` because the memory values aren't the same - so the equal test fails, so probably need to filter out all memory entries from this comparison tests - as those can be different. But this is a minor issue.

Reply:
I think there can be a last separate log with the memory metrics, different from the one the logs the other metrics. There is already a last log with the total FLOS for instance after training (which probably could be combined with the one for memory usage).

For the test, it should just add the flag `skip_memory_metrics` to not add those logs.

Reply:
> I think there can be a last separate log with the memory metrics, different from the one that logs the other metrics. There is already a last log with the total FLOS for instance after training (which probably could be combined with the one for memory usage).

yes, but that's too early, if you're referring to:

https://github.com/huggingface/transformers/blob/master/src/transformers/trainer.py#L1057-L1060

since there is a callback afterward, so the measurement needs to occur just before return.
 
> For the test, it should just add the flag `skip_memory_metrics` to not add those logs (and we should test for that flag in the Trainer before logging of course).

Well, I can easily test both ways with the `CaptureStd` helper - but I hear you - I will add a test.

Reply:
full test added

Reply:
Ok, so then we can add the logs at the end of train/evaluate as a separate event I think.

Reply:
I'm not quite sure what exactly you're suggesting here - would you please use some pseudo-code to exemplify?

I'm asking because `logs()` calls `on_logs` callback, so it can't just be moved as it'd change the order of callbacks.

Reply:
Yes, but we could have a last log event when the metrics are reported. In any case we can already merge this and see if users want the additional log before changing everything (since this would impact quite a few tests probably).

Reply:
Sounds good then. Let's start using it and I'm sure we will see some nuances we haven't thought of yet. Thank you, @sgugger 
!